python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,501 @@
+
+
+@pytest.fixture(scope="session")
+def output_python_type_func_fixture():
+    # This fixture helps to check the response
+    # when the function return value is not an Arrow
+    # defined data type. Instead here the returned value
+    # is of type int in Python.
+    def const_return(ctx, scalar):
+        return 42
+
+    func_name = "test_output_type"
+    in_types = {"array": pa.int64()}
+    out_type = pa.int64()
+    doc = {
+        "summary": "add function scalar",
+        "description": "add function"
+    }
+    pc.register_scalar_function(const_return, func_name, doc,
+                                in_types, out_type)
+    return const_return, func_name
+
+
+@pytest.fixture(scope="session")
+def varargs_check_func_fixture():
+    def varargs_check(ctx, *values):
+        base_val = values[:2]
+        res = pc.call_function("add", base_val)
+        for other_val in values[2:]:
+            res = pc.call_function("add", [res, other_val])
+        return res
+    func_name = "test_varargs_function"
+    in_types = {"array1": pa.int64(),
+                "array2": pa.int64(),
+                }
+    doc = {"summary": "n add function",
+           "description": "add N number of arrays"
+           }
+    pc.register_scalar_function(varargs_check, func_name, doc,
+                                in_types, pa.int64())
+
+    return varargs_check, func_name
+
+
+@pytest.fixture(scope="session")
+def raise_func_fixture():
+    def raise_func(ctx):
+        raise ValueError("Test function with raise")
+    func_name = "test_raise"
+    doc = {
+        "summary": "test function with raise",
+        "description": "function with a raise"
+    }
+    pc.register_scalar_function(raise_func, func_name, doc,
+                                {}, pa.int64())
+    return raise_func, func_name
+
+
+def check_scalar_function(func_fixture,
+                          input,
+                          run_in_dataset=True,
+                          batch_length=global_batch_length):
+    function, name = func_fixture
+    expected_output = function(mock_udf_context(batch_length), *input)
+    func = pc.get_function(name)
+    assert func.name == name
+
+    result = pc.call_function(name, input)
+
+    assert result == expected_output
+    if run_in_dataset:
+        field_names = [f'field{index}' for index, in_arr in input]
+        table = pa.Table.from_arrays(input, field_names)
+        dataset = ds.dataset(table)
+        func_args = [ds.field(field_name) for field_name in field_names]
+        result_table = dataset.to_table(
+            columns={'result': ds.field('')._call(name, func_args)})
+        assert result_table.column(0).chunks[0] == expected_output
+
+
+def test_scalar_udf_array_unary(unary_func_fixture):
+    check_scalar_function(unary_func_fixture,
+                          [
+                              pa.array([10, 20], pa.int64())
+                          ],
+                          mock_udf_context()
```

Comment:
Hmm, wait. Why is this being passed as the third argument to `check_scalar_function`? The types don't match.

Reply:
Ah yes, this is not required. Removing it.